Add support for parseable-output straight from the frontend

artemcm opened the pull request:
This PR adds initial support for emitting parseable-output JSON directly from `swift-frontend`.
"began" and "finished" messages are currently supported.

- The existing parseable-output library used by the Driver is sunk into `Basic` and extended with entry-points that accept a `CompilerInvocation` of a `swift-frontend`, in addition to existing entry-points used by the Driver that accept a `Job`. 
- A new diagnostic consumer is added in order to capture all produced diagnostics during the frontend's execution, mirroring what the printing consumer would do, except simply accumulating diagnostics, on a per-primary-input basis.
- For batch jobs, "began" and "finished" messages are emitted on a per-primary-input basis, with quasi-PIDs assigned to the messages of jobs we are "faking". The same scheme as used by the driver is employed to generate quasi-PIDs: start at -1000 and decrement for each next primary input. Non-batch jobs get a real OS PID. 

Note: The code in ParseableOutput.h and ParseableOutput.cpp with a  `Job`-based interface is not new. I tried moving these files carefully to preserve history, but enough stuff seems to have changed to not get nice diffs here.

artemcm committed:
Sink the parseable-output library into `Basic` from `Driver`

artemcm committed:
Add support for emitting parseable-outpuit "began" message from swift-frontend

artemcm committed:
Add support for emitting parseable-output "finished" message and per-primary messages for batched jobs
Inside swift-frontend

artemcm committed:
Implement frontend parseable-output batch job quasi-PID assignment.
Starting at a crude -1000, each invocation primary input will get its own unique quasi-Pid.
Invocations with only one primary (non-batch) will get a real OS Pid.

The selection of the constant starting point matches what the driver does when outputting its parseable output.

artemcm committed:
Suppress the output of the PrintingDiagnosticConsumer when in Frontend Parseable-Output mode

artemcm committed:
Add frontend parseable output tests

artemcm committed:
Match diagnostic output formatting from the PrintDiagnosticConsumer

brentdax commented on include/swift/Basic/ParseableOutput.h:
```diff
@@ -0,0 +1,72 @@
+/// PIDs (which are always positive). We start at -1000 here as a crude but
+/// harmless hedge against colliding with an errno value that might slip
+/// into the stream of real PIDs (say, due to a TaskQueue bug).
+const int QUASI_PID_START = -1000;
```
Before I dig into the implementation details: When the driver assigns quasi-PIDs, it knows what all of the frontend jobs are doing and can prevent two of them from using the same quasi-PID. How are we preventing this if the frontend jobs themselves are assigning the quasi-PIDs?

(If you aren’t doing anything to handle this, one solution might be to use a different scheme. For instance, you could negate the real PID, multiply it by the batch size the driver was trying to assign to each job, and then add the index of the source file you’re currently building.)

artemcm replied:
You're correct, there is currently nothing to coordinate this or avoid collisions among different frontend jobs, yet. I like your suggested scheme, let me think about this a little bit.

artemcm commented:
@swift-ci Please Test Source Compatibility

artemcm commented:
@swift-ci Please Smoke Test

artemcm commented:
@swift-ci Please Smoke Test

artemcm commented:
@swift-ci Please Smoke Test Windows platform

artemcm commented:
@swift-ci Please Smoke Test Windows platform

artemcm commented:
@swift-ci Please Smoke Test Windows platform

artemcm commented:
@swift-ci Please Smoke Test

artemcm commented:
@swift-ci Please Smoke Test

artemcm commented:
@swift-ci Please Smoke Test Linux Platform

artemcm commented:
@swift-ci Please Smoke Test

artemcm commented:
@swift-ci Please smoke Test

artemcm commented:
@swift-ci Please Test

artemcm commented:
@swift-ci Please Test Source Compatibility

brentdax commented on lib/Basic/ParseableOutput.cpp:
```diff
@@ -0,0 +1,484 @@
+    const auto &supplementaryOutputPaths =
+        primarySpecificFiles.SupplementaryOutputs;
+    supplementaryOutputPaths.forEachSetOutput([&](const std::string &output) {
+      Outputs.push_back(OutputPair(file_types::lookupTypeForExtension(
+                                       llvm::sys::path::extension(output)),
+                                   output));
```
I'm not sure it's a good idea to use `lookupTypeForExtension()` like this. A build system could pass filenames with no extension or the wrong extension; if they do, we should still treat these correctly, using a file type based on which field in the supplementary output path field we found the path in. (This probably means `forEachSetOutput` would pass the file type to the lambda.)

artemcm replied:
This is a great observation, thank you. We do indeed have all the information we need to determines this based on action and known file type instead of inferring it. I modified `forAllOutputPaths` to pass along the filetype.

brentdax commented on lib/Basic/ParseableOutput.cpp:
```diff
@@ -0,0 +1,484 @@
+
+    // Output for this Primary
+    auto OutputFile = PrimaryInput.outputFilename();
+    Outputs.push_back(OutputPair(file_types::lookupTypeForExtension(
```
Like below, I think you could avoid using `lookupTypeForExtension()` here. Would `FrontendOptions::formatForPrincipalOutputFileForAction()` get the right file type?

brentdax commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -2811,6 +2831,22 @@ int swift::performFrontend(ArrayRef<const char *> Args,
+    // If we got this far, we need to suppress the output of the
+    // PrintingDiagnosticConsumer to ensure that only the parseable-output
+    // is emitted
+    PDC.setSuppressOutput(true);
```
Just checking: We're disabling the PrintingDiagnosticConsumer instead of never adding it in the first place because we need the PDC while we're parsing arguments, including the argument that sets this option?

artemcm replied:
Exactly. In fact, it seems that this is why 'setSuppressOutput' exists in the first place.

artemcm commented on include/swift/Basic/SupplementaryOutputPaths.h:
```diff
@@ -171,6 +172,42 @@ struct SupplementaryOutputPaths {
+      fn(LdAddCFilePath); 
+    if (!ModuleSummaryOutputPath.empty())
+      fn(ModuleSummaryOutputPath); 
+  }
```
I could not agree more. 
And _yes_ on the `FrontendOptions::forAllOutputPaths()`, which suffers from the same problem/smell however. I will use `forAllOutputPaths` for now, but add a TODO for myself and a FIXME in the code to capture that this is less-than-deal.

artemcm commented on include/swift/Frontend/FrontendInputsAndOutputs.h:
```diff
@@ -104,6 +104,11 @@ class FrontendInputsAndOutputs {
+  /// Iterates over primary inputs, exposing their unique ordered index
+  /// If \p fn returns true, exit early and return true.
+  bool forEachPrimaryInputWithIndex(
+      llvm::function_ref<bool(const InputFile &, unsigned index)> fn) const;
```
There are a bunch of existing clients and wrappers for `forEachPrimaryInput`, none of which seem to care about the index. It seemed cleaner to me to not force on those clients the fact that behind the scenes `FrontendInputsAndOutputs` maintains this unique ordered index, when not necessary.

artemcm commented on include/swift/Frontend/FrontendOptions.h:
```diff
@@ -183,6 +183,11 @@ class FrontendOptions {
 
+  /// Emit parseable-output directly from the frontend, instead of relying
+  /// the driver to emit it. This is used in context where frontend jobs are executed by
+  /// clients other than the driver.
```
Made the explanation clearer, thank you.

brentdax replied:
It's much clearer now!

brentdax commented on lib/Frontend/FrontendOptions.cpp:
```diff
@@ -223,11 +223,23 @@ void FrontendOptions::forAllOutputPaths(
                                   &outs.ModuleInterfaceOutputPath,
                                   &outs.PrivateModuleInterfaceOutputPath,
                                   &outs.ObjCHeaderOutputPath,
+                                  &outs.DependenciesFilePath,
```
Be careful here—the existing client of `forAllOutputPaths()` is emitting dependency files, and I could imagine that having dependency files refer to themselves might be a problem.

(Perhaps that client can filter on the file type, once you're passing a file type.)

brentdax commented on lib/Frontend/FrontendInputsAndOutputs.cpp:
```diff
@@ -379,6 +379,15 @@ void FrontendInputsAndOutputs::forEachOutputFilename(
+    llvm::function_ref<void(StringRef, file_types::ID)> fn) const {
+  (void)forEachInputProducingAMainOutputFile(
+      [&](const InputFile &input) -> bool {
+        fn(input.outputFilename(), input.getType());
```
Could be wrong, but I think `getType()` here gets the type of the input file, not the type of the output file.

artemcm commented on include/swift/Basic/ParseableOutput.h:
```diff
@@ -0,0 +1,72 @@
+#include "swift/Basic/LLVM.h"
+#include "swift/Basic/TaskQueue.h"
+#include "swift/Driver/Job.h"
+#include "swift/Frontend/Frontend.h"
```
Thank you for pointing this out.
The implementation's `Message` types already lended themselves pretty nicely to being separated in this fashion with some core basics being kept in a `Basic` library. The specific portions have now been split into their new homes in `Driver` and `Frontend`.

artemcm commented:
@brentdax Thank you for such thorough and insightful feedback.

artemcm commented:
@swift-ci please test

artemcm commented:
@swift-ci please test

artemcm commented:
@swift-ci please test

artemcm commented:
@swift-ci please test Windows platform

artemcm commented:
@swift-ci please test Windows platform

artemcm commented:
@swift-ci please test Windows platform

artemcm commented:
@swift-ci please test Windows platform

artemcm commented:
@swift-ci please test Windows platform

artemcm commented:
@swift-ci please test

artemcm committed:
Add host-specific include for `getpid()` in FrontendTool.

artemcm commented:
@swift-ci please test

artemcm commented:
@swift-ci please test macOS platform

artemcm commented:
@swift-ci please test

artemcm committed:
Remove unnecessary platform-specific checks from tests

artemcm commented:
@swift-ci please test Windows platform

artemcm commented:
@swift-ci please test

artemcm committed:
Remove unnecessary XFAILs

artemcm commented:
@swift-ci please test

artemcm commented:
@swift-ci please test

artemcm commented:
@brentdax ping. The feedback above has been addressed, I believe.

artemcm committed:
Refactor Basic/Parseable-Output to remove dependency on Driver & Frontend
Introducing new entry-points that can be used from both Driver and Frontend clients, using an intermediary new type: `DetailedMessagePayload`, when needed.

artemcm commented:
@swift-ci please test

artemcm commented:
@swift-ci please test macOS platform

brentdax commented on include/swift/Basic/ParseableOutput.h:
```diff
@@ -0,0 +1,76 @@
+
+using OutputPair = std::pair<file_types::ID, std::string>;
+
+struct DetailedMessagePayload {
+  std::string Executable;
+  SmallVector<std::string, 16> Arguments;
+  std::string CommandLine;
+  SmallVector<CommandInput, 4> Inputs;
+  SmallVector<OutputPair, 8> Outputs;
+};
```
This type ought to have a doc comment.

I'd also like you to think carefully about its name. Yes, from the perspective of the `emit*Message()` functions, this is the payload of the `DetailedMessage` you're emitting. But that's an implementation detail of this subsystem.

A better type name would come from asking, "what is this type to the *clients* of this header"? I might call it something like a "job description" or "process info" or maybe a "quasi-invocation" or something. I'll leave the exact decision to you.

artemcm replied:
Thank you for the insight, I fully agree. 
I have renamed the type and it now has a doc to describe its purpose.

beccadax reviewed (changes requested):
Much better! I have a couple nits to pick, but this has the right shape now.

artemcm commented:
@swift-ci please test

artemcm commented:
@swift-ci please test
